coderd/prometheusmetrics/aggregator.go

```diff
@@ -0,0 +1,248 @@
+		Name:      "metrics_aggregator_execution_cleanup_seconds",
+		Help:      "Histogram for duration of metrics aggregator cleanup in seconds.",
+		Buckets:   []float64{0.001, 0.005, 0.010, 0.025, 0.050, 0.100, 0.500, 1, 5, 10, 30},
+	})
+	err = registerer.Register(cleanupHistogram)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricsAggregator{
+		log:                    logger,
+		metricsCleanupInterval: metricsCleanupInterval,
+
+		collectCh: make(chan (chan<- prometheus.Metric), sizeCollectCh),
+		updateCh:  make(chan updateRequest, sizeUpdateCh),
+
+		updateHistogram:  updateHistogram,
+		cleanupHistogram: cleanupHistogram,
+	}, nil
+}
+
+func (ma *MetricsAggregator) Run(ctx context.Context) func() {
+	ctx, cancelFunc := context.WithCancel(ctx)
+	done := make(chan struct{})
+
+	cleanupTicker := time.NewTicker(ma.metricsCleanupInterval)
+	go func() {
+		defer close(done)
+		defer cleanupTicker.Stop()
+
+		for {
+			select {
+			case req := <-ma.updateCh:
+				ma.log.Debug(ctx, "metrics aggregator: update metrics")
+
+				timer := prometheus.NewTimer(ma.updateHistogram)
+			UpdateLoop:
+				for _, m := range req.metrics {
+					for i, q := range ma.queue {
+						if q.username == req.username && q.workspaceName == req.workspaceName && q.agentName == req.agentName && q.Name == m.Name {
+							ma.queue[i].AgentMetric.Value = m.Value
+							ma.queue[i].expiryDate = req.timestamp.Add(ma.metricsCleanupInterval)
+							continue UpdateLoop
+						}
+					}
+
+					ma.queue = append(ma.queue, annotatedMetric{
+						username:      req.username,
+						workspaceName: req.workspaceName,
+						agentName:     req.agentName,
+
+						AgentMetric: m,
+
+						expiryDate: req.timestamp.Add(ma.metricsCleanupInterval),
+					})
+				}
+
+				timer.ObserveDuration()
+			case inputCh := <-ma.collectCh:
+				ma.log.Debug(ctx, "metrics aggregator: collect metrics")
+
+				for _, m := range ma.queue {
+					desc := prometheus.NewDesc(m.Name, metricHelpForAgent, agentMetricsLabels, nil)
+					valueType, err := asPrometheusValueType(m.Type)
+					if err != nil {
+						ma.log.Error(ctx, "can't convert Prometheus value type", slog.F("name", m.Name), slog.F("type", m.Type), slog.F("value", m.Value), slog.Error(err))
+						continue
+					}
+					constMetric := prometheus.MustNewConstMetric(desc, valueType, m.Value, m.username, m.workspaceName, m.agentName)
+					inputCh <- constMetric
+				}
+				close(inputCh)
+			case <-cleanupTicker.C:
+				ma.log.Debug(ctx, "metrics aggregator: clean expired metrics")
+
+				timer := prometheus.NewTimer(ma.cleanupHistogram)
+
+				now := time.Now()
+
+				var hasExpiredMetrics bool
+				for _, m := range ma.queue {
+					if now.After(m.expiryDate) {
+						hasExpiredMetrics = true
+						break
+					}
+				}
+
+				if hasExpiredMetrics {
+					var j int
+					fresh := make([]annotatedMetric, 0, len(ma.queue))
+					for _, m := range ma.queue {
+						if m.expiryDate.After(now) {
+							fresh[j] = m
+							j++
```

Comment:
```suggestion
					fresh := make([]annotatedMetric, 0, len(ma.queue))
					for _, m := range ma.queue {
						if m.expiryDate.After(now) {
							fresh = append(fresh, m)
```
I probably explained the make change poorly as this was also a required part of it. A go-ish simplification, if you will. You can do this or then revert the change because assigning `fresh[j]` with `make(..., 0)` would lead to a bad time. 😄

Reply:
I trust your skills, thanks 👍